cdc/log.cc

```diff
@@ -1289,7 +1290,11 @@ class transformer final {
         auto partition_slice = query::partition_slice(std::move(bounds), std::move(static_columns), std::move(regular_columns), std::move(opts));
         auto command = ::make_lw_shared<query::read_command>(_schema->id(), _schema->version(), partition_slice, row_limit);
 
-        return _ctx._proxy.query(_schema, std::move(command), std::move(partition_ranges), cl, service::storage_proxy::coordinator_query_options(default_timeout(), empty_service_permit(), client_state)).then(
+        const auto select_cl = write_cl >= db::consistency_level::QUORUM
```

Comment:
@piodul found a bug here. Apparently we can't compare consistency_level with < and >. The order of values in the enum is not semantically correct:
```
enum class consistency_level {
    ANY, MIN_VALUE = ANY,
    ONE,
    TWO,
    THREE,
    QUORUM,
    ALL,
    LOCAL_QUORUM,
    EACH_QUORUM,
    SERIAL,
    LOCAL_SERIAL,
    LOCAL_ONE, MAX_VALUE = LOCAL_ONE
};
```

LOCAL_ONE is after QUORUM

Reply:
Also I the condition should look like this:
`std::min(QUORUM, std::max(ONE, write_cl))`
with a semantically correct comparison. This means if write_cl is LOCAL_ONE, TWO or THREE, we should select with LOCAL_ONE, TWO or THREE as well

Reply:
I think we will have to implement some sort of comparator for CL but the problem is the order is not total and depends on RF. For example with RF=1 order is:
ANY, ONE, LOCAL_ONE, TWO, QUORUM, SERIAL, LOCAL_QUORUM, LOCAL_SERIAL, EACH_QUORUM, THREE, ALL
while for RF=6 the order is:
ANY, ONE, LOCAL_ONE, TWO, THREE, QUORUM, SERIAL, LOCAL_QUORUM, LOCAL_SERIAL, EACH_QUORUM, ALL

so the comparator will have to take RF into account.

Reply:
@kostja Could you please confirm the placement of SERIAL and LOCAL_SERIAL in the order?

Reply:
I'm also not sure whether QUORUM < LOCAL_QUORUM or the other way around. QUORUM will require more ACKs but it will probably be easier to achieve QUORUM than LOCAL_QUORUM because the number of replicas for QUORUM is bigger than number of replicas for LOCAL_QUORUM.

Reply:
If we define strength of the consistency level as the number of replicas needed for read/write, then this function might come in handy:

```
size_t db::block_for(keyspace& ks, consistency_level cl);
```

It's defined in `db/consistency_level.cc`. It takes RF of the keyspace into consideration.

Reply:
I saw it but it won't be useful because ONE, LOCAL_ONE and ANY are equal there. It is useful to copy some stuff from it like `ks.get_replication_strategy().get_replication_factor()` and `quorum_for(ks)` though

Reply:
Nice catch, I missed LOCAL_ONE and deliberately squeezed TWO and THREE into ONE (maybe wrongly, but they are flaky with their guarantees of all or nothing, depending on RF). I'll update.

Speaking of (LOCAL_)SERIAL - I need guidance here.

Reply:
SERIAL is QUORUM, LOCAL_SERIAL is LOCAL_QUORUM. Treat them this way please.

Reply:
QUORUM and LOCAL_QUORUM are not comparable at all. Imagine you have a cluster of 6 nodes, 3 nodes in dc1 and 3 in dc2. 
If 2 nodes out of 3 in DC1 are down, QUORUM will pass, LOCAL_QUORUM in DC1 will not.

Imagine you have 3 nodes in DC1 down. LOCAL_QUORUM ind DC2 will pass, QUORUM will not.

Reply:
Maybe we can go with a simple solution ANY,ONE, LOCAL_ONE, TWO, THREE -> ONE and other CL -> QUORUM?

Reply:
Because creating comparator clearly makes no sense.